File: src/cli/auth/logout.ts
import type { Command } from "commander";
import {
  clearApiKey,
  clearAll,
  getDefaultWorkspace,
  getWorkspaces,
  removeWorkspace,
} from "../../lib/config.ts";
import { printJson } from "../../lib/output.ts";

export function registerLogout(auth: Command): void {
  auth
    .command("logout")
    .description("Clear stored credentials")
    .option("--all", "Remove all workspaces (default: only active workspace)")
    .action((opts: { all?: boolean }) => {
      if (opts.all) {
        clearAll();
        printJson({ ok: true, cleared: "all" });
        return;
      }
      const defaultWs = getDefaultWorkspace();
      if (defaultWs) {
        try {
          removeWorkspace(defaultWs);
        } catch {
          // workspace already gone
        }
      }
      clearApiKey();
      const newDefault = getDefaultWorkspace();
      printJson({
        ok: true,
        removed: defaultWs ?? null,
        remaining_workspaces: Object.keys(getWorkspaces()),
        default_workspace: newDefault ?? null,
      });
    });
}

File: internal/cli/auth/logout.go
package auth

import (
	"sort"

	"github.com/spf13/cobra"

	"github.com/shhac/lin/internal/config"
	"github.com/shhac/lin/internal/output"
)

func registerLogout(auth *cobra.Command) {
	var all bool

	cmd := &cobra.Command{
		Use:   "logout",
		Short: "Clear stored credentials",
		Args:  cobra.NoArgs,
		Run: func(_ *cobra.Command, _ []string) {
			if all {
				if err := config.ClearAll(); err != nil {
					output.PrintError(err.Error())
					return
				}
				output.PrintJSON(map[string]any{"ok": true, "cleared": "all"})
				return
			}

			defaultWs := config.GetDefaultWorkspace()
			if defaultWs != "" {
				_ = config.RemoveWorkspace(defaultWs)
			}
			_ = config.ClearApiKey()

			newDefault := config.GetDefaultWorkspace()
			remaining := sortedKeys(config.GetWorkspaces())

			result := map[string]any{
				"ok":                   true,
				"removed":             nilIfEmpty(defaultWs),
				"remaining_workspaces": remaining,
				"default_workspace":   nilIfEmpty(newDefault),
			}
			output.PrintJSON(result)
		},
	}

	cmd.Flags().BoolVar(&all, "all", false, "Remove all workspaces (default: only active workspace)")
	auth.AddCommand(cmd)
}

func sortedKeys(m map[string]config.Workspace) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

func nilIfEmpty(s string) any {
	if s == "" {
		return nil
	}
	return s
}
